File: src/main/java/com/demonwav/mcdev/creator/CanaryProjectSettingsWizard.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.creator;

import com.demonwav.mcdev.asset.PlatformAssets;
import com.demonwav.mcdev.platform.PlatformType;
import com.demonwav.mcdev.platform.canary.CanaryProjectConfiguration;

import com.intellij.openapi.options.ConfigurationException;
import org.apache.commons.lang.WordUtils;
import org.jetbrains.annotations.NotNull;

import javax.swing.JComboBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextField;

public class CanaryProjectSettingsWizard extends MinecraftModuleWizardStep {

    private JTextField pluginNameField;
    private JTextField pluginVersionField;
    private JTextField mainClassField;
    private JPanel panel;
    private JTextField authorsField;
    private JComboBox loadOrderBox;
    private JTextField dependField;
    private JLabel title;
    private JComboBox<String> canaryVersionBox;

    private CanaryProjectConfiguration settings;
    private final MinecraftProjectCreator creator;

    public CanaryProjectSettingsWizard(@NotNull MinecraftProjectCreator creator) {
        this.creator = creator;
    }

    @Override
    public JComponent getComponent() {
        settings = (CanaryProjectConfiguration) creator.getSettings().get(PlatformType.CANARY);
        if (settings == null) {
            return panel;
        }

        String name = WordUtils.capitalize(creator.getArtifactId());
        pluginNameField.setText(name);
        pluginVersionField.setText(creator.getVersion());

        if (settings != null && !settings.isFirst) {
            pluginNameField.setEditable(false);
            pluginVersionField.setEditable(false);
        }

        mainClassField.setText(creator.getGroupId().toLowerCase() + '.' + creator.getArtifactId().toLowerCase()
            + '.' + name);

        if (creator.getSettings().size() > 1) {
            mainClassField.setText(mainClassField.getText() + settings.type.getNormalName());
        }

        switch (settings.type) {
            case CANARY:
                title.setIcon(PlatformAssets.CANARY_ICON_2X);
                title.setText("<html><font size=\"5\">Canary Settings</font></html>");
                break;
            case NEPTUNE:
                title.setIcon(PlatformAssets.NEPTUNE_ICON_2X);
                title.setText("<html><font size=\"5\">Neptune Settings</font></html>");
                break;
            default:
        }

        return panel;
    }

    @Override
    public boolean isStepVisible() {
        settings = (CanaryProjectConfiguration) creator.getSettings().get(PlatformType.CANARY);
        return settings != null;
    }

    @Override
    public boolean validate() throws ConfigurationException {
        return validate(pluginNameField, pluginVersionField, mainClassField, authorsField, dependField, pattern);
    }

    @Override
    public void onStepLeaving() {
        this.settings.pluginName = pluginNameField.getText();
        this.settings.pluginVersion = pluginVersionField.getText();
        this.settings.mainClass = mainClassField.getText();
        this.settings.setAuthors(this.authorsField.getText());
        this.settings.enableEarly = this.loadOrderBox.getSelectedIndex() == 0 ? false : true;
        this.settings.setDependencies(this.dependField.getText());
        this.settings.canaryVersion = (String) canaryVersionBox.getSelectedItem();
    }

    @Override
    public void updateDataModel() {}
}

File: src/main/kotlin/com/demonwav/mcdev/creator/CanaryProjectSettingsWizard.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2018 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.creator

import com.demonwav.mcdev.asset.PlatformAssets
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.canary.CanaryProjectConfiguration
import org.apache.commons.lang.WordUtils
import javax.swing.JComboBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField

class CanaryProjectSettingsWizard(private val creator: MinecraftProjectCreator) : MinecraftModuleWizardStep() {

    private lateinit var pluginNameField: JTextField
    private lateinit var pluginVersionField: JTextField
    private lateinit var mainClassField: JTextField
    private lateinit var panel: JPanel
    private lateinit var authorsField: JTextField
    private lateinit var loadOrderBox: JComboBox<*>
    private lateinit var dependField: JTextField
    private lateinit var title: JLabel
    private lateinit var canaryVersionBox: JComboBox<String>
    private lateinit var errorLabel: JLabel

    private var settings: CanaryProjectConfiguration? = null

    override fun getComponent(): JComponent {
        settings = creator.settings[PlatformType.CANARY] as? CanaryProjectConfiguration
        if (settings == null) {
            return panel
        }

        val name = WordUtils.capitalize(creator.artifactId.replace('-', ' '))
        pluginNameField.text = name
        pluginVersionField.text = creator.version

        if (settings != null && !settings!!.isFirst) {
            pluginNameField.isEditable = false
            pluginVersionField.isEditable = false
        }

        mainClassField.text = creator.groupId.replace("-", "").toLowerCase() + "." +
            creator.artifactId.replace("-", "").toLowerCase() + "." + name.replace(" ", "")

        if (creator.settings.size > 1) {
            mainClassField.text = mainClassField.text + settings!!.type.normalName
        }

        when (settings!!.type) {
            PlatformType.CANARY -> {
                title.icon = PlatformAssets.CANARY_ICON_2X
                title.text = "<html><font size=\"5\">Canary Settings</font></html>"
            }
            PlatformType.NEPTUNE -> {
                title.icon = PlatformAssets.NEPTUNE_ICON_2X
                title.text = "<html><font size=\"5\">Neptune Settings</font></html>"
            }
            else -> {}
        }

        getVersionSelector(settings!!.type)
            .onSuccess { it.set(canaryVersionBox) }
            .onError { errorLabel.isVisible = true }

        return panel
    }

    override fun isStepVisible(): Boolean {
        settings = creator.settings[PlatformType.CANARY] as? CanaryProjectConfiguration
        return settings != null
    }

    override fun validate(): Boolean {
        return validate(pluginNameField, pluginVersionField, mainClassField, authorsField, dependField, MinecraftModuleWizardStep.pattern) &&
            canaryVersionBox.selectedItem != null
    }

    override fun onStepLeaving() {
        this.settings!!.pluginName = pluginNameField.text
        this.settings!!.pluginVersion = pluginVersionField.text
        this.settings!!.mainClass = mainClassField.text
        this.settings!!.setAuthors(this.authorsField.text)
        this.settings!!.isEnableEarly = this.loadOrderBox.selectedIndex != 0
        this.settings!!.setDependencies(this.dependField.text)
        this.settings!!.canaryVersion = canaryVersionBox.selectedItem as? String ?: ""
    }

    override fun updateDataModel() {}
}
